### PR TITLE
Make torch.arange consistent with numpy.arange

### LINKED ISSUE #5556: arange(0.1, 0.5, 0.05) returns wrong results
```python
import torch, numpy as np

torch.arange(0.1, 0.50, 0.05)
# 0.1000
# 0.1500
# 0.2000
# 0.2500
# 0.3000
# 0.3500
# 0.4000
#[torch.FloatTensor of size 7]


np.arange(0.1, 0.50, 0.05, dtype = np.float32)
#array([0.1       , 0.15      , 0.20000002, 0.25      , 0.3       ,
#       0.35000002, 0.4       , 0.45000002], dtype=float32)
```

0.45 is missing from PyTorch output, which seems strange, 0.45 is clearly less than 0.50, even considering floating point errror.

on `0.4.0a0+1fdb392`

I guess the cause is that `arange` computes number of elements by `floor((end - start) / step)` and float error impacts this calculation, but the departure from NumPy behavior in simple cases like this (not even https://github.com/pytorch/pytorch/issues/4877) may be confusing.

### PR DESCRIPTION
**Pull request by li-roy**
This should fix #5556. I couldn't reproduce on CUDA, but I changed that as well anyways.

```
>>> torch.arange(0.1, 0.50, 0.05)

 0.1000
 0.1500
 0.2000
 0.2500
 0.3000
 0.3500
 0.4000
 0.4500
[torch.FloatTensor of size 8]
```

### CODE REVIEW

File: aten/src/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -2828,14 +2828,20 @@ void THTensor_(range)(THTensor *r_, accreal xmin, accreal xmax, accreal step)
...
-#endif
-  if (m)
-    xmax -= step;
-  THTensor_(range)(r_, xmin, xmax, step);
+  ptrdiff_t size;
+  real i = 0;
+
+  THArgCheck(step > 0 || step < 0, 3, "step must be a non-null number");
+  THArgCheck(((step > 0) && (xmax >= xmin)) || ((step < 0) && (xmax <= xmin))
+              , 2, "upper bound and larger bound incoherent with step sign");
```
Comment (goldsborough) [inline aten/src/TH/generic/THTensorMath.c]: Nit: I think you mean "inconsistent"

File: aten/src/TH/generic/THTensorMath.c [hunk 2]
```diff
@@ -2828,14 +2828,20 @@ void THTensor_(range)(THTensor *r_, accreal xmin, accreal xmax, accreal step)
...
-#else
-  int m = (xmax - xmin) % step == 0;
-#endif
-  if (m)
-    xmax -= step;
-  THTensor_(range)(r_, xmin, xmax, step);
+  ptrdiff_t size;
+  real i = 0;
+
+  THArgCheck(step > 0 || step < 0, 3, "step must be a non-null number");
```
Comment (goldsborough) [inline aten/src/TH/generic/THTensorMath.c]: Is there a numeric reason why `step != 0` doesn't work?

File: aten/src/TH/generic/THTensorMath.c [hunk 3]
```diff
@@ -2828,14 +2828,20 @@ void THTensor_(range)(THTensor *r_, accreal xmin, accreal xmax, accreal step)
...
+  THArgCheck(((step > 0) && (xmax >= xmin)) || ((step < 0) && (xmax <= xmin))
+              , 2, "upper bound and larger bound incoherent with step sign");
+
+  size = (ptrdiff_t) ceil((xmax - xmin) / step);
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: I don't think it's a fix for numerical errors. This seems to implement a different algorithm (if we were to interpret it using the real numbers without rounding errors). Previously we did `floor((xmax - xmin) / step)`, now we do `ceil((xmax - xmin) / step)`

**Comment by ezyang** [comment]
@li-roy In #5608 @Officium introduces a much simpler numerical fix. What do you think of that patch?

**Comment by colesbury** [comment]
A few comments:

1) More tests please. We've consistently had buggy implementations of `arange`. One outcome of this is that we should be convinced of the correctness of `arange` as opposed to just squashing bugs as they're reported. For short lists, explicitly check the actual value, e.g. `assertEqual(arange(0, 1), [0])`. For larger values, at least check the first and last elements along with the size. 

2) The behavior should be consistent with NumPy's implementation. I think someone posted a link to it in #5608. This is perhaps more important than improving numerical accuracy.

**Comment by li-roy** [comment]
@ezyang Yeah, i rewrote arange instead of changing range cause I think fmod is actually causing error in some cases.

Also, I looked at the numpy implementation and it looks like nmp_ceil is just calls C99 ceil, which I decided to use here.

File: aten/src/TH/generic/THTensorMath.c (hunk 2 above)
Comment (li-roy) [reply aten/src/TH/generic/THTensorMath.c]: I dont think it really makes sense because if xmin and xmax are different, we can't really step from xmin to xmax with a step of x. If xmin and xmax are the same, how would we choose a size?

**Comment by li-roy** [comment]
This implementation should match what numpy does. Numpy actually has the same floating point problem in arange, but this pr should make it so pytorch is consistent with that. An example of where numpy has this error as well:

this pr pytorch
```
>>> torch.arange(0.6, 0.9, 0.1)

 0.6000
 0.7000
 0.8000
 0.9000
[torch.FloatTensor of size 4]
```

numpy
```
>>> np.arange(0.6, 0.9, 0.1)
array([0.6, 0.7, 0.8, 0.9])
```

File: aten/src/TH/generic/THTensorMath.c (hunk 2 above)
Comment (colesbury) [reply aten/src/TH/generic/THTensorMath.c]: I think Peter was suggesting that the line read:

```
THArgCheck(step != 0, 3, ...);
```

(not that we allow a zero step)

The error message "step must be a non-null number" is a bit awkward. (I think something like "step must not be 0" will be better).

If you're concerned about `nan`, maybe just add an explicit check for it.

(I realize that this is taken from the `range` code above)

File: aten/src/THC/generic/THCTensorMath.cu [hunk 4]
```diff
@@ -491,14 +491,20 @@ void THCTensor_(range)(THCState *state, THCTensor *r_, accreal xmin, accreal xma
...
-  if (m)
-    xmax -= step;
-  THCTensor_(range)(state, r_, xmin, xmax, step);
+  THCAssertSameGPU(THCTensor_(checkGPU)(state, 1, r_));
+  THArgCheck(step > 0 || step < 0, 3, "step must be a non-null number");
+  THArgCheck(((step > 0) && (xmax >= xmin)) || ((step < 0) && (xmax <= xmin))
+              , 2, "upper bound and larger bound inconsistent with step sign");
+  ptrdiff_t size = (ptrdiff_t) ceil(ScalarConvert<accreal, double>::to(xmax - xmin) / step);
+  if (THCTensor_(nElement)(state, r_) != size) THCTensor_(resize1d)(state, r_, size);
+  THCTensor *r = THCTensor_(isContiguous)(state, r_)
```
Comment (colesbury) [inline aten/src/THC/generic/THCTensorMath.cu]: This should just be:

```
THCTensor *r = THCTensor_(newContiguous)(state, r_);
```

The pattern is that newContiguous just `increfs` and returns `r_` if the tensor is already contiguous.

File: aten/src/THC/generic/THCTensorMath.cu [hunk 5]
```diff
@@ -491,14 +491,20 @@ void THCTensor_(range)(THCState *state, THCTensor *r_, accreal xmin, accreal xma
...
+                 ? r_
+                 : THCTensor_(newContiguous)(state, r_);
+  LinspaceOp<real,accreal> linspace_method(xmin, step);
+  thrust::device_ptr<real> data_(THCTensor_(data)(state, r));
+  thrust::tabulate(data_, data_ + size, linspace_method);
+  if (!THCTensor_(isContiguous)(state, r_)) THCTensor_(freeCopyTo)(state, r, r_);
```
Comment (colesbury) [inline aten/src/THC/generic/THCTensorMath.cu]: And this should just be:

```
THCTensor_(freeCopyTo)(state, r, r_);
```

**Review by colesbury — COMMENTED** [review]
Nice! Thanks for fixing this. A few comments inline:

File: aten/src/TH/generic/THTensorMath.c (hunk 2 above)
Comment (li-roy) [reply aten/src/TH/generic/THTensorMath.c]: Ah okay, that makes sense now. Sorry for the misunderstanding, Peter.

If nobody has any strong objections, I'm going leave this as `step > 0 || step < 0`. I don't think it makes too much of a difference in terms of readability.

**Review by colesbury — APPROVED** [review]
lgtm
